Zoom out mode: scale iframe instead of contents

ellatrix opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

This PR moves the scaling from zoom out mode from the iframe body to the frame element.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

We should avoid adding styles such as margin, height and positioning _inside_ the iframe as it might interfere with theme styles.

See also #46929.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

One important detail: I refactored the in between inserter by moving the width/height calculation inside the popover ref and adding a new popover prop "overlay" to tell it to overlay this reference.

The problem with the current approach is that when the width is calculated in the in between inserted, it isn't scaled.

In any case, I find this new approach a bit clearer, it clears up some in between inserter logic.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

Enable the zoomed out mode experiment. Go to the site editor and toggle it. Ensure popovers are positioned correctly and scrolling works as expected.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

ellatrix committed:
Zoom out mode: scale iframe instead of contents

ellatrix committed:
Refactor in between inserter dimensions

ellatrix commented on packages/block-editor/src/components/block-popover/inbetween.js:
```diff
@@ -158,21 +118,32 @@ function BlockPopoverInbetween( {
 					}
 				}
 
-				return new window.DOMRect( left, top, 0, 0 );
```
The style calculation above is now absorbed in the popover ref calculation.

ellatrix committed:
Reduce diff

youknowriad commented on packages/components/src/popover/types.ts:
```diff
@@ -148,6 +148,13 @@ export type PopoverProps = {
+	 *
+	 * @default false
+	 */
+	overlay?: boolean;
```
Isn't that the same as the "placement" prop. In other words, should this be another "placement" instead of a separate prop?

ellatrix replied:
It could be, if it's fine to add an extra option to the floating UI options.

ciampo replied:
I agree that having `overlay` as a separate prop would basically "override" whatever behavior was set through `placement`.

Probably adding an extra option to `placement` feels like the correct choice here, even if that means deviating from `@floating-ui`. The best thing here would be to make these changes directly in `@floating-ui`, so that we can keep using `placement` without any changes in `@wordpress/components`

jsnajdr replied:
If I understand it correctly, `overlay` is a type of placement where the popover completely covers the anchor element: it's placed at the anchor's top-left corner, and resized to exactly match the size of the anchor.

`@floating-ui`'s job is then to maintain the placement when things move, resize and scroll.

It could be a nice contribution to `@floating-ui`, it makes sense to me.

ciampo replied:
I agree with @jsnajdr , the best solution here is to keep `overlay` as an additonal value for the `placement` prop and remove the separate `overlay` prop.

talldan commented:
> The problem with the current approach is that when the width is calculated in the in between inserter, it isn't scaled.

What's the specific blocker to making the Inbetween inserter component do the scaling? (as an alternative to adding the new prop to Popover)

ellatrix commented:
> What's the specific blocker to making the Inbetween inserter component do the scaling? (as an alternative to adding the new prop to Popover)

The in between inserter then needs to be aware of the scaling, which is not great. It means some duplicate logic in checking the iframe element for the scale transform. I preferred this approach.

ciampo commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -244,6 +245,20 @@ function Iframe( {
+						: props.style?.marginTop,
+					marginBottom: scale
+						? -( ( contentHeight * ( 1 - scale ) ) / 2 ) + frameSize
+						: props.style?.marginBottom,
```
We could maybe add some code comments to explain the math here?

ciampo commented on packages/components/src/popover/utils.ts:
```diff
@@ -157,6 +157,15 @@ export const getFrameOffset = (
+		return 1;
+	}
+	const transform = frameElement.style.transform;
+	return parseFloat( transform.replace( /scale\((\d+\.?\d*)\)/, '$1' ) );
```
If kept within `Popover`, this logic should be made more robust.

- For example, what happens if the popover is nested inside two or more scaled iframes?

- The regex doesn't cover all possible ways of setting a `scale` value:

  - the `scale` function can have more than one argument (ie. `scale(x, y)`)
  - the style could be set via a `matrix` syntax
  - ...and maybe other edge cases that I'm ignoring

Probably the best way would be to `getComputedStyles` and derive the scale from it.

Also, we should take into account that the `iframe` could be scaled with different values for `x` and `y` axis.

ciampo reviewed (commented):
Thank you for working on this, Ella!

Since `Popover` is already quite a complex component which has had (and still has) a few quirks especially when [dealing with cross-document references](https://github.com/floating-ui/floating-ui/pull/2043) (and more specifically [around scaling](https://github.com/floating-ui/floating-ui/issues/1594)), I had the same initial thoughts as @talldan 

> What's the specific blocker to making the Inbetween inserter component do the scaling? (as an alternative to adding the new prop to Popover)

The scaling seems to be a problem quite specific to the inbetween inserter at the moment, and therefore it seemed fitting to me that the complexity would stay in that component. But I can also see why you'd add the logic directly to `Popover`.

A few more points:

- it would be great if we could add some tests, although I'm not sure there's an easy way to do so (for example, `floating-ui` uses e2e + visual regression on the screenshots)
- the _best_ solution, in my opinion, would be to add this feature directly to `floating-ui`
- finally, would it be possible to move the changes to the `Popover` component to a separate 

cc'ing also @jsnajdr who's been recently taking a look at `Popover` and `@floating-ui`

jsnajdr commented on packages/components/src/popover/index.tsx:
```diff
@@ -270,6 +273,14 @@ const UnforwardedPopover = (
+						return rects.reference;
+					},
+			  }
+			: undefined,
```
Seems to me that the `overlay` option makes all the other middlewares pointless, except maybe `frameOffset`.

`offset`, `flip`, `size`, `shift`, `arrow`: they all make sense only for popovers attached to one side of the anchor. For a popover that fully covers the anchor, none of them is useful.

Therefore, the `middleware` should be more like:
```js
const middleware = overlay
  ? [ /* array of overlay middlewares */ ]
  : [ /* array of classic middlewares */ ];
```

ellatrix replied:
I didn't want to make any assumptions here. The iframe middleware is still needed though. Theoretically you could also still shift the position I guess.

ellatrix committed:
Address feedback

ellatrix commented:
> the best solution, in my opinion, would be to add this feature directly to floating-ui

Happy to try that once we merge it here.

I think I addressed all feedback now.

ciampo commented on packages/components/src/popover/index.tsx:
```diff
@@ -361,7 +365,10 @@ const UnforwardedPopover = (
+		placement:
+			( normalizedPlacementFromProps === 'overlay'
+				? undefined
+				: normalizedPlacementFromProps ) || 'bottom',
```
The `<Popover />`'s `placement` prop already has a default value of `bottom-start` (see props destructuring at the top of the component), and otherwise `floating-ui` defaults to `bottom` for its [`placement` option](https://floating-ui.com/docs/computePosition#placement) — I think we can safely delete the `bottom` fallback here

ciampo reviewed (commented):
Hey @ellatrix , sorry for the delay in my reply — still catching up with a long queue after my last AFK.

> Happy to try that once we merge it here.

Sounds good to me — I'd love to see us contributing back to `@floating-ui` as a follow-up to this PR.

Could you please also add an entry to the components package's [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md)?

Thank you! 🙏

ellatrix committed:
Address feedback

ellatrix commented:
@ciampo Thanks! I adjusted the change log and addressed your feedback.

ciampo commented on packages/components/CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@
 -   `Dropdown`: deprecate `position`  prop, use `popoverProps` instead ([46865](https://github.com/WordPress/gutenberg/pull/46865)).
 -   `Button`: improve padding for buttons with icon and text. ([46764](https://github.com/WordPress/gutenberg/pull/46764)).
 -   `ColorPalette`: Use computed color when css variable is passed to `ColorPicker` ([47181](https://github.com/WordPress/gutenberg/pull/47181)).
+-   `Popover`: add `overlay` option to the `placement` prop.
```
```suggestion
-   `Popover`: add `overlay` option to the `placement` prop ([47004](https://github.com/WordPress/gutenberg/pull/47004)).
```

ciampo commented on packages/components/CHANGELOG.md:
```diff
@@ -22,6 +23,7 @@
  ([#47384](https://github.com/WordPress/gutenberg/pull/47384)).
 -   `Button`: Convert to TypeScript ([#46997](https://github.com/WordPress/gutenberg/pull/46997)).
 -   `QueryControls`: Convert to TypeScript ([#46721](https://github.com/WordPress/gutenberg/pull/46721)).
+-   `Popover`: Take iframe element scaling into account.
```
```suggestion
-   `Popover`: Take iframe element scaling into account ([47004](https://github.com/WordPress/gutenberg/pull/47004)).
```

ciampo reviewed (commented):
Thank you Ella! I had a good look at the component in Storybook too, and couldn't find any major issues.

<details>
<summary>One change that I made was to enable the Storybook examples to showcase the new `overlay` value for `placement` — would you be able to apply this change as well?</summary>

```diff
diff --git a/packages/components/src/popover/stories/index.tsx b/packages/components/src/popover/stories/index.tsx
index f5a8fab45d..15723becf1 100644
--- a/packages/components/src/popover/stories/index.tsx
+++ b/packages/components/src/popover/stories/index.tsx
@@ -31,6 +31,7 @@ const AVAILABLE_PLACEMENTS: PopoverProps[ 'placement' ][] = [
 	'left',
 	'left-start',
 	'left-end',
+	'overlay',
 ];
 
 const meta: ComponentMeta< typeof Popover > = {
@@ -170,7 +171,12 @@ export const AllPlacements: ComponentStory< typeof Popover > = ( {
 		</h2>
 		<div>
 			{ AVAILABLE_PLACEMENTS.map( ( p ) => (
-				<PopoverWithAnchor key={ p } placement={ p } { ...args }>
+				<PopoverWithAnchor
+					key={ p }
+					placement={ p }
+					{ ...args }
+					resize={ p === 'overlay' ? true : args.resize }
+				>
 					{ children }
 					<div>
 						<small>(placement: { p })</small>
```
</details>

Another aspect that came to mind while playing with the component in Storybook, is that when the `placement` is set to `'overlay'` a few props may behave in a weird / unexpected way, or become ineffective: `offset`, `noArrow`, `flip`....

Is the behaviour of the component when tweaking those props  respecting the consumer's expectations? Should we change anything? Should we at least document those quirks?

ellatrix committed:
Update packages/components/CHANGELOG.md

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

ellatrix committed:
Update packages/components/CHANGELOG.md

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

ellatrix committed:
Address feedback

jsnajdr commented:
> Another aspect that came to mind while playing with the component in Storybook, is that when the placement is set to 'overlay' a few props may behave in a weird / unexpected way, or become ineffective: offset, noArrow, flip....

If there are two sets of props that are mutually incompatible (I also pointed it out in https://github.com/WordPress/gutenberg/pull/47004#discussion_r1071262999), maybe that's a hint we should have two components, not just one?

One of them would be the old `<Popover>`, without the `overlay` prop, and the new one, I'll jokingly call it `<Layover>`, would have a limited set of props that really matter. And they would be both implemented in terms of the current `<Popover>`, which would become private. One with `overlay={ false }`, the other with `overlay={ true }`.

ellatrix commented:
> Another aspect that came to mind while playing with the component in Storybook, is that when the `placement` is set to `'overlay'` a few props may behave in a weird / unexpected way, or become ineffective: `offset`, `noArrow`, `flip`....
> 
> Is the behaviour of the component when tweaking those props respecting the consumer's expectations? Should we change anything? Should we at least document those quirks?

`offset` seems to work as expected. Tested both a positive and negative offset. `flip` and `shift` have no effect, right. I guess they could have an effect if you don't resize the popover?

> If there are two sets of props that are mutually incompatible

I don't think these props are incompatible. Some props may just not have any effect in some circumstances. I would leave things as they are now and see how it evolves.

jsnajdr commented:
> `flip` and `shift` have no effect, right. I guess they could have an effect if you don't resize the popover?

They both make sense for popovers that are _outside_ the anchor element, and they move the popover if there's not enough outer space in the preferred location. I don't see what they could possibly do for an overlay popover.

Other than these API details, the PR seems OK and mergeable to me.

ellatrix commented:
> I don't see what they could possibly do for an overlay popover.

I guess an overlay position could still overflow the anchor if it isn't resized. Anyway, we can figure out these details as use cases pop up.

Thanks for the reviews!

ciampo commented:
Opened https://github.com/floating-ui/floating-ui/issues/2171 to get the conversation started about implementing the overlay functionality directly in `floating-ui`

ciampo commented:
Update: the `floating-ui` authors decided against supporting `overlay` placement (or a similar feature) directly in the library — https://github.com/floating-ui/floating-ui/issues/2171#issuecomment-1425292789

ellatrix commented:
Interesting. It's pretty much like their `center` example. Makes me wonder if we should rename it to that. Btw, I think the resizing should be optional and done in a separate prop?

ciampo commented:
`center` to me suggests centering reference and floating, but without affecting the floating size

IMO, if we want to keep the current `overlay` behaviour within `Popover`, (ie. the floating element matching both coordinates AND size of the reference element), `overlay` sounds more fitting as a name.

Otherwise, if we wanted to expose a `center` placement, we'd somehow need to allow the consumer to specify that the want the floating to be resized as well.

----

Also — this conversation around `center` reminded me that there are a few values for the legacy `position` prop that we couldn't map to `placement` exactly because it didn't offer a `center` option:

https://github.com/WordPress/gutenberg/blob/8dee3e6764da100b20ad9abd41af5df086b5f42f/packages/components/src/popover/utils.ts#L63-L70

We could also consider exposing a `center` option for `placement`, separately from `overlay`?

Mamaduka commented:
The changes in this PR introduced the following warning when switching the device types in the Site Editor.

```
Warning: Removing a style property during rerender (margin) when a conflicting property is set (marginBottom) can lead to styling bugs. To avoid this, don't mix shorthand and non-shorthand properties for the same value; instead, replace the shorthand with separate values.
```

## Screenshot
![CleanShot 2023-04-18 at 15 14 50](https://user-images.githubusercontent.com/240569/232760541-ba280849-0d89-4872-9a27-77bdd858f1f9.png)

ciampo commented:
> The changes in this PR introduced the following warning when switching the device types in the Site Editor.
> 
> ```
> Warning: Removing a style property during rerender (margin) when a conflicting property is set (marginBottom) can lead to styling bugs. To avoid this, don't mix shorthand and non-shorthand properties for the same value; instead, replace the shorthand with separate values.
> ```
> 
> ## Screenshot
> ![CleanShot 2023-04-18 at 15 14 50](https://user-images.githubusercontent.com/240569/232760541-ba280849-0d89-4872-9a27-77bdd858f1f9.png)

@ellatrix , would you be able to take a look at that?

ciampo commented:
> The changes in this PR introduced the following warning when switching the device types in the Site Editor.
> 
> ```
> Warning: Removing a style property during rerender (margin) when a conflicting property is set (marginBottom) can lead to styling bugs. To avoid this, don't mix shorthand and non-shorthand properties for the same value; instead, replace the shorthand with separate values.
> ```
> 
> ## Screenshot
> ![CleanShot 2023-04-18 at 15 14 50](https://user-images.githubusercontent.com/240569/232760541-ba280849-0d89-4872-9a27-77bdd858f1f9.png)

@Mamaduka this should have been fixed by #50441
